File: realingo_app/lib/tech_services/rest/rest_api.dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' as http;
import 'package:realingo_app/model/program.dart';
import 'package:realingo_app/tech_services/app_config.dart';
import 'package:realingo_app/tech_services/rest/rest_data.dart';
import 'package:realingo_app/tech_services/result.dart';

/*
Rest API wrapper
 */
class RestApi {
  static const String _restApiBaseUrl = '${AppConfig.apiUrl}/api/v0';

  static Future<Result<T>> _runGet<T>(String request, T Function(dynamic decodedBody) bodyToResult,
      {dynamic Function(http.Response response)? decodeResponse}) async {
    try {
      http.Response response = await http.get(Uri.parse('$_restApiBaseUrl/$request')).timeout(Duration(seconds: 10));

      if (response.statusCode != 200) {
        return Result.ko(AppError.RestRequestFailed);
      }

      dynamic jsonDecoded;
      if (decodeResponse == null) {
        jsonDecoded = json.decode(response.body);
      } else {
        jsonDecoded = decodeResponse(response);
      }

      T result = bodyToResult(jsonDecoded);
      return Result.ok(result);
    } on TimeoutException catch (_) {
      return Result.ko(AppError.RestRequestFailed);
    }
  }

  static Future<Result<List<Language>>> getAvailableOriginLanguages(String learnedLanguageUri) async {
    return _runGet('available_origin_languages?learned_language_uri=$learnedLanguageUri', (dynamic decodedBody) {
      final languages = List<Language>.unmodifiable((decodedBody as List<dynamic>)
          .map((dynamic i) => RestLanguage.fromJson(i as Map<String, dynamic>))
          .map<Language>((e) => Language(e.uri, e.label)));
      return languages;
    });
  }

  static Future<Result<List<Language>>> getAvailableLearnedLanguages() async {
    return _runGet('available_learned_languages', (dynamic decodedBody) {
      final languages = List<Language>.unmodifiable((decodedBody as List<dynamic>)
          .map((dynamic i) => RestLanguage.fromJson(i as Map<String, dynamic>))
          .map<Language>((e) => Language(e.uri, e.label)));

      return languages;
    });
  }

  static Future<Result<LearningProgram>> getProgramByLanguage(Language learnedLanguage, Language originLanguage) async {
    return _runGet(
        'program_by_language?learned_language_uri=${learnedLanguage.uri}&origin_language_uri=${originLanguage.uri}',
        (dynamic decodedBody) {
      return _programFromRest(decodedBody);
    });
  }

  static Future<Result<LearningProgram>> getProgram(String programUri) async {
    return _runGet('program?program_uri=$programUri', (dynamic decodedBody) {
      return _programFromRest(decodedBody);
    });
  }

  static LearningProgram _programFromRest(dynamic responseBody) {
    final restProgram = RestLearningProgram.fromJson(responseBody as Map<String, dynamic>);

    final lessons = List<LessonInProgram>.unmodifiable(
        restProgram.lessons.map<LessonInProgram>((e) => LessonInProgram(e.uri, e.label, e.description)));

    return LearningProgram(restProgram.uri, restProgram.learnedLanguageUri, restProgram.originLanguageUri, lessons);
  }

  static Future<Result<Lesson>> getLesson(String programUri, String lessonUri) async {
    return _runGet('lesson?program_uri=$programUri&lesson_uri=$lessonUri', (dynamic decodedBody) {
      final restLesson = RestLesson.fromJson(decodedBody as Map<String, dynamic>);

      final exercises = List<Exercise>.unmodifiable(restLesson.exercises.map<Exercise>((e) => Exercise(
          e.uri,
          fromRest(e.exerciseType),
          Sentence(
              e.sentence.uri,
              e.sentence.sentence,
              e.sentence.translation,
              e.sentence.hint,
              List<ItemInSentence>.unmodifiable(e.sentence.items.map<ItemInSentence>((e) => ItemInSentence(
                  e.startIndex,
                  e.endIndex,
                  e.label,
                  List<ItemTranslation>.unmodifiable(e.translations
                      .map<ItemTranslation>((e) => ItemTranslation(e.translation, e.englishDefinition))))))))));

      return Lesson(restLesson.uri, restLesson.label, restLesson.description, exercises);
    });
  }

  static ExerciseType fromRest(RestExerciseType exerciseType) {
    switch (exerciseType) {
      case RestExerciseType.TranslateToLearningLanguage:
        return ExerciseType.TranslateToLearningLanguage;
      case RestExerciseType.Repeat:
        return ExerciseType.Repeat;
    }
  }

  static Future<Result<Uint8List>> getRecord(String languageUri, String sentence) async {
    return _runGet('sentence_record?language_uri=$languageUri&sentence=$sentence',
        (dynamic decodedBody) => decodedBody as Uint8List,
        decodeResponse: (http.Response response) => response.bodyBytes);
  }
}

File: back/src/main/kotlin/co/globers/realingo/back/restapi/v0/RestApi.kt
package co.globers.realingo.back.restapi.v0

import co.globers.realingo.back.model.Language
import co.globers.realingo.back.services.ProgramCache
import co.globers.realingo.back.services.ProgramKey
import co.globers.realingo.back.services.TextToSpeech
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController



// variable to suppress on this deprecate
@Suppress("DEPRECATION")
private const val jsonUtf8Header: String = MediaType.APPLICATION_JSON_UTF8_VALUE

@RestController
class RestApi(val programCache: ProgramCache, val textToSpeech: TextToSpeech) {

    @GetMapping("/api/v0/available_origin_languages", produces = [jsonUtf8Header])
    suspend fun getAvailableOriginLanguages(
            @RequestParam(value = "learned_language_uri") learnedLanguageUri: String): List<RestLanguage> {
        return programCache.availableOriginLanguages.map { ModelToRestAdapter.toRest(it) }
    }

    @GetMapping("/api/v0/available_learned_languages", produces = [jsonUtf8Header])
    suspend fun getAvailableLearnedLanguages(): List<RestLanguage> {
        return programCache.availableLearnedLanguages.map { ModelToRestAdapter.toRest(it) }
    }

    // dart client use latin1 format if utf-8 not explicitly specified by the server
    // https://github.com/dart-lang/http/issues/175
    @GetMapping("/api/v0/program_by_language", produces = [jsonUtf8Header])
    suspend fun getProgramByLanguage(
            @RequestParam(value = "learned_language_uri") learnedLanguageUri: String,
            @RequestParam(value = "origin_language_uri") originLanguageUri: String): RestLearningProgram {

        val originLanguage = Language.fromUri(originLanguageUri)
        val learnedLanguage = Language.fromUri(learnedLanguageUri)

        val programUri = programCache.availablePrograms.getValue(ProgramKey(learnedLanguage, originLanguage))
        val program = programCache.programsByUri.getValue(programUri)
        val restProgram = ModelToRestAdapter.toRest(program)
        return restProgram
    }

    @GetMapping("/api/v0/program", produces = [jsonUtf8Header])
    suspend fun getProgram(
        @RequestParam(value = "program_uri") programUri: String): RestLearningProgram {

        val program = programCache.programsByUri.getValue(programUri)
        val restProgram = ModelToRestAdapter.toRest(program)
        return restProgram
    }


    @GetMapping("/api/v0/lesson", produces = [jsonUtf8Header])
    suspend fun getLesson(
        @RequestParam(value = "program_uri") programUri: String,
        @RequestParam(value = "lesson_uri") lessonUri: String
    ): RestLesson {

        val program = programCache.programsByUri.getValue(programUri)
        val lesson = program.lessons.first { it.uri == lessonUri }
        val restLesson = ModelToRestAdapter.toRest(lesson)
        return restLesson
    }

    @GetMapping("/api/v0/sentence_record", produces = [MediaType.APPLICATION_OCTET_STREAM_VALUE])
    suspend fun getSentenceRecord(
        @RequestParam(value = "language_uri") languageUri: String,
        @RequestParam(value = "sentence") sentence: String) : ByteArray {

        val result = textToSpeech.getRecord(Language.fromUri(languageUri), sentence)
        return result
    }


}
